src/olympia/devhub/tasks.py

```diff
@@ -139,22 +132,56 @@ def create_version_for_upload(addon, upload, channel):
         add_dynamic_theme_tag(version)
 
 
+@task
+def create_initial_validation_results():
+    """Returns the initial validation results for the next tasks in the
+    validation chain. Should only be called directly by Validator."""
+    results = deepcopy(amo.VALIDATOR_SKELETON_RESULTS)
+    return results
+
+
 def validation_task(fn):
-    """Wrap a validation task so that it runs with the correct flags, then
-    parse and annotate the results before returning."""
+    """Wrap a validation task so that it runs with the correct flags and
+    handles errors (mainly because Celery's error handling does not work for
+    us).
+
+    ALL the validation tasks but `create_initial_validation_results()` should
+    use this decorator. Tasks decorated with `@validation_task` should have the
+    following declaration:
 
-    @task(bind=True, ignore_result=False,  # We want to pass the results down.
+        @validation_task
+        def my_task(results, pk):
+            # ...
+            return results
+
+    Notes:
+
+    * `results` is automagically passed to each task in the validation chain
+      and created by `create_initial_validation_results()` at the beginning of
+      the chain. It MUST be the first argument of the task.
+    * `pk` is passed to each task when added to the validation chain.
+    * the validation chain is defined in the `Validator` class.
+    """
+    @task(bind=True,
+          ignore_result=False,  # We want to pass the results down.
           soft_time_limit=settings.VALIDATOR_TIMEOUT)
+    @use_primary_db
     @wraps(fn)
-    def wrapper(task, id_or_path, *args, **kwargs):
-        # This is necessary to prevent timeout exceptions from being set
-        # as our result, and replacing the partial validation results we'd
-        # prefer to return.
+    def wrapper(task, results, pk, *args, **kwargs):
+        # This is necessary to prevent timeout exceptions from being set as our
+        # result, and replacing the partial validation results we'd prefer to
+        # return.
         task.ignore_result = True
         try:
-            data = fn(id_or_path, **kwargs)
-            results = json.loads(force_text(data))
-            return results
+            # All validation tasks should receive `results`.
+            if not results:
+                raise Exception('Unexpected call to a validation task without '
+                                '`results`')
```

Comment:
Maybe `assert` instead, and/or write a test for this.

Reply:
added a test case